amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -463,10 +465,10 @@ private static boolean hashRangeOfShardsIsComplete(@NonNull List<Shard> shards)
     static boolean checkIfDescendantAndAddNewLeasesForAncestors(final String shardId,
             final InitialPositionInStreamExtended initialPosition, final Set<String> shardIdsOfCurrentLeases,
             final Map<String, Shard> shardIdToShardMapOfAllKinesisShards,
```

Comment:
does shardIdToShardMap contains all the shards of the stream or the shards with respect to the initial position of the stream ?

Ex: Stream with 1000 shards and have 300 days of retention can contain 1 million shards. It takes approximately 1000 pages to complete list shards. Does this code avoid that 1000 pages of list shards ? 
If the KCL change is made to call ListShards @ TIMESTAMP, @ AT_TRIMHORIZON or AT_LATEST we will not have shards at multiple hierarchy

or Is this check used only in the recovery mode when there are holes detected ?

Reply:
This check is only used in recovery mode when there are holes detected, at least in KCL 2.x. In KCL 1.x it could also be exercised during PeriodicShardSync but that will be handled in KCL 1.x impl.